04_labs/corpus/external_frameworks/pandoc/pandoc-cli/man/pandoc.1
.\" Automatically generated by Pandoc 3.9
.\"
.TH "pandoc" "1" "2025\-02\-02" "pandoc 3.9" "Pandoc User\[cq]s Guide"
.SH NAME
pandoc - general markup converter
.SH SYNOPSIS
\f[CR]pandoc\f[R] [\f[I]options\f[R]] [\f[I]input\-file\f[R]]\&...
.SH DESCRIPTION
Pandoc is a Haskell library for converting from one markup format to
another, and a command\-line tool that uses this library.
.PP
Pandoc can convert between numerous markup and word processing formats,
including, but not limited to, various flavors of Markdown, HTML, LaTeX
and Word docx.
For the full lists of input and output formats, see the
\f[CR]\-\-from\f[R] and \f[CR]\-\-to\f[R] options below.
Pandoc can also produce PDF output: see creating a PDF, below.
.PP
Pandoc\(cqs enhanced version of Markdown includes syntax for tables,
definition lists, metadata blocks, footnotes, citations, math, and much
more.
See below under Pandoc\(cqs Markdown.
.PP
Pandoc has a modular design: it consists of a set of readers, which
parse text in a given format and produce a native representation of the
document (an \f[I]abstract syntax tree\f[R] or AST), and a set of
writers, which convert this native representation into a target format.
Thus, adding an input or output format requires only adding a reader or
writer.
Users can also run custom pandoc filters to modify the intermediate AST.
.PP
Because pandoc\(cqs intermediate representation of a document is less
expressive than many of the formats it converts between, one should not
expect perfect conversions between every format and every other.
Pandoc attempts to preserve the structural elements of a document, but
not formatting details such as margin size.
And some document elements, such as complex tables, may not fit into
pandoc\(cqs simple document model.
While conversions from pandoc\(cqs Markdown to all formats aspire to be
perfect, conversions from formats more expressive than pandoc\(cqs
Markdown can be expected to be lossy.
.SS Using pandoc
If no \f[I]input\-files\f[R] are specified, input is read from
\f[I]stdin\f[R].
Output goes to \f[I]stdout\f[R] by default.
For output to a file, use the \f[CR]\-o\f[R]/\f[CR]\-\-output\f[R]
option:
.IP
.EX
pandoc \-o output.html input.txt
.EE
.PP
By default, pandoc produces a document fragment.
To produce a standalone document (e.g.\ a valid HTML file including
\f[CR]<head>\f[R] and \f[CR]<body>\f[R]), use the \f[CR]\-s\f[R] or
\f[CR]\-\-standalone\f[R] flag:
.IP
.EX
pandoc \-s \-o output.html input.txt
.EE
.PP
For more information on how standalone documents are produced, see
Templates below.
.PP
If multiple input files are given, pandoc will concatenate them all
(with blank lines between them) before parsing.
(Use \f[CR]\-\-file\-scope\f[R] to parse files individually.)
.SS Specifying formats
The format of the input and output can be specified explicitly using
command\-line options.
The input format can be specified using the \f[CR]\-f/\-\-from\f[R]
option, the output format using the \f[CR]\-t/\-\-to\f[R] option.
Thus, to convert \f[CR]hello.txt\f[R] from Markdown to LaTeX, you could
type:
.IP
.EX
pandoc \-f markdown \-t latex hello.txt
.EE
.PP
To convert \f[CR]hello.html\f[R] from HTML to Markdown:
.IP
.EX
pandoc \-f html \-t markdown hello.html
.EE
.PP
Supported input and output formats are listed below under Options (see
\f[CR]\-f\f[R] for input formats and \f[CR]\-t\f[R] for output formats).
You can also use \f[CR]pandoc \-\-list\-input\-formats\f[R] and
\f[CR]pandoc \-\-list\-output\-formats\f[R] to print lists of supported
formats.
.PP
If the input or output format is not specified explicitly, pandoc will
attempt to guess it from the extensions of the filenames.
Thus, for example,
.IP
.EX
pandoc \-o hello.tex hello.txt
.EE
.PP
will convert \f[CR]hello.txt\f[R] from Markdown to LaTeX.
If no output file is specified (so that output goes to
\f[I]stdout\f[R]), or if the output file\(cqs extension is unknown, the
output format will default to HTML.
If no input file is specified (so that input comes from
\f[I]stdin\f[R]), or if the input files\(cq extensions are unknown, the
input format will be assumed to be Markdown.
.SS Character encoding
Pandoc uses the UTF\-8 character encoding for both input and output.
If your local character encoding is not UTF\-8, you should pipe input
and output through \f[CR]iconv\f[R]:
.IP
.EX
iconv \-t utf\-8 input.txt | pandoc | iconv \-f utf\-8
.EE
.PP
Note that in some output formats (such as HTML, LaTeX, ConTeXt, RTF,
OPML, DocBook, and Texinfo), information about the character encoding is
included in the document header, which will only be included if you use
the \f[CR]\-s/\-\-standalone\f[R] option.
.SS Creating a PDF
To produce a PDF, specify an output file with a \f[CR].pdf\f[R]
extension:
.IP
.EX
pandoc test.txt \-o test.pdf
.EE
.PP
By default, pandoc will use LaTeX to create the PDF, which requires that
a LaTeX engine be installed (see \f[CR]\-\-pdf\-engine\f[R] below).
Alternatively, pandoc can use ConTeXt, roff ms, or HTML as an
intermediate format.
To do this, specify an output file with a \f[CR].pdf\f[R] extension, as
before, but add the \f[CR]\-\-pdf\-engine\f[R] option or
\f[CR]\-t context\f[R], \f[CR]\-t html\f[R], or \f[CR]\-t ms\f[R] to the
command line.
The tool used to generate the PDF from the intermediate format may be
specified using \f[CR]\-\-pdf\-engine\f[R].
.PP
You can control the PDF style using variables, depending on the
intermediate format used: see variables for LaTeX, variables for
ConTeXt, variables for \f[CR]wkhtmltopdf\f[R], variables for ms.
When HTML is used as an intermediate format, the output can be styled
using \f[CR]\-\-css\f[R].
.PP
To debug the PDF creation, it can be useful to look at the intermediate
representation: instead of \f[CR]\-o test.pdf\f[R], use for example
\f[CR]\-s \-o test.tex\f[R] to output the generated LaTeX.
You can then test it with \f[CR]pdflatex test.tex\f[R].
.PP
When using LaTeX, the following packages need to be available (they are
included with all recent versions of TeX Live): \f[CR]amsfonts\f[R],
\f[CR]amsmath\f[R], \f[CR]lm\f[R], \f[CR]unicode\-math\f[R],
\f[CR]iftex\f[R], \f[CR]listings\f[R] (if the \f[CR]\-\-listings\f[R]
option is used), \f[CR]fancyvrb\f[R], \f[CR]longtable\f[R],
\f[CR]booktabs\f[R], \f[CR]multirow\f[R] (if the document contains a
table with cells that cross multiple rows), \f[CR]graphicx\f[R] (if the
document contains images), \f[CR]bookmark\f[R], \f[CR]xcolor\f[R],
\f[CR]soul\f[R], \f[CR]geometry\f[R] (with the \f[CR]geometry\f[R]
variable set), \f[CR]setspace\f[R] (with \f[CR]linestretch\f[R]), and
\f[CR]babel\f[R] (with \f[CR]lang\f[R]).
If \f[CR]CJKmainfont\f[R] is set, \f[CR]xeCJK\f[R] is needed if
\f[CR]xelatex\f[R] is used, else \f[CR]luatexja\f[R] is needed if
\f[CR]lualatex\f[R] is used.
\f[CR]framed\f[R] is required if code is highlighted in a scheme that
use a colored background.
The use of \f[CR]xelatex\f[R] or \f[CR]lualatex\f[R] as the PDF engine
requires \f[CR]fontspec\f[R].
\f[CR]lualatex\f[R] uses \f[CR]selnolig\f[R] and \f[CR]lua\-ul\f[R].
\f[CR]xelatex\f[R] uses \f[CR]bidi\f[R] (with the \f[CR]dir\f[R]
variable set).
If the \f[CR]mathspec\f[R] variable is set, \f[CR]xelatex\f[R] will use
\f[CR]mathspec\f[R] instead of \f[CR]unicode\-math\f[R].
The \f[CR]csquotes\f[R] package will be used for typography if the
\f[CR]csquotes\f[R] variable or metadata field is set to a true value.
The \f[CR]natbib\f[R], \f[CR]biblatex\f[R], \f[CR]bibtex\f[R], and
\f[CR]biber\f[R] packages can optionally be used for citation rendering.
If math with \f[CR]\(rscancel\f[R], \f[CR]\(rsbcancel\f[R], or
\f[CR]\(rsxcancel\f[R] is used, the \f[CR]cancel\f[R] package is needed.
The following packages will be used to improve output quality if
present, but pandoc does not require them to be present:
\f[CR]upquote\f[R] (for straight quotes in verbatim environments),
\f[CR]microtype\f[R] (for better spacing adjustments),
\f[CR]parskip\f[R] (for better inter\-paragraph spaces), \f[CR]xurl\f[R]
(for better line breaks in URLs), and \f[CR]footnotehyper\f[R] or
\f[CR]footnote\f[R] (to allow footnotes in tables).
.SS Reading from the Web
Instead of an input file, an absolute URI may be given.
In this case pandoc will fetch the content using HTTP:
.IP
.EX
pandoc \-f html \-t markdown https://www.fsf.org
.EE
.PP
It is possible to supply a custom User\-Agent string or other header
when requesting a document from a URL:
.IP
.EX
pandoc \-f html \-t markdown \-\-request\-header User\-Agent:\(dqMozilla/5.0\(dq \(rs
  https://www.fsf.org
.EE
.SH OPTIONS
.SS General options
.TP
\f[CR]\-f\f[R] \f[I]FORMAT\f[R], \f[CR]\-r\f[R] \f[I]FORMAT\f[R], \f[CR]\-\-from=\f[R]\f[I]FORMAT\f[R], \f[CR]\-\-read=\f[R]\f[I]FORMAT\f[R]
Specify input format.
\f[I]FORMAT\f[R] can be:
.RS
.IP \(bu 2
\f[CR]asciidoc\f[R] (AsciiDoc markup)
.IP \(bu 2
\f[CR]bibtex\f[R] (BibTeX bibliography)
.IP \(bu 2
\f[CR]biblatex\f[R] (BibLaTeX bibliography)
.IP \(bu 2
\f[CR]bits\f[R] (BITS XML, alias for \f[CR]jats\f[R])
.IP \(bu 2
\f[CR]commonmark\f[R] (CommonMark Markdown)
.IP \(bu 2
\f[CR]commonmark_x\f[R] (CommonMark Markdown with extensions)
.IP \(bu 2
\f[CR]creole\f[R] (Creole 1.0)
.IP \(bu 2
\f[CR]csljson\f[R] (CSL JSON bibliography)
.IP \(bu 2
\f[CR]csv\f[R] (CSV table)
.IP \(bu 2
\f[CR]tsv\f[R] (TSV table)
.IP \(bu 2
\f[CR]djot\f[R] (Djot markup)
.IP \(bu 2
\f[CR]docbook\f[R] (DocBook)
.IP \(bu 2
\f[CR]docx\f[R] (Word docx)
.IP \(bu 2
\f[CR]dokuwiki\f[R] (DokuWiki markup)
.IP \(bu 2
\f[CR]endnotexml\f[R] (EndNote XML bibliography)
.IP \(bu 2
\f[CR]epub\f[R] (EPUB)
.IP \(bu 2
\f[CR]fb2\f[R] (FictionBook2 e\-book)
.IP \(bu 2
\f[CR]gfm\f[R] (GitHub\-Flavored Markdown), or the deprecated and less
accurate \f[CR]markdown_github\f[R]; use \f[CR]markdown_github\f[R] only
if you need extensions not supported in \f[CR]gfm\f[R].
.IP \(bu 2
\f[CR]haddock\f[R] (Haddock markup)
.IP \(bu 2
\f[CR]html\f[R] (HTML)
.IP \(bu 2
\f[CR]ipynb\f[R] (Jupyter notebook)
.IP \(bu 2
\f[CR]jats\f[R] (JATS XML)
.IP \(bu 2
\f[CR]jira\f[R] (Jira/Confluence wiki markup)
.IP \(bu 2
\f[CR]json\f[R] (JSON version of native AST)
.IP \(bu 2
\f[CR]latex\f[R] (LaTeX)
.IP \(bu 2
\f[CR]markdown\f[R] (Pandoc\(cqs Markdown)
.IP \(bu 2
\f[CR]markdown_mmd\f[R] (MultiMarkdown)
.IP \(bu 2
\f[CR]markdown_phpextra\f[R] (PHP Markdown Extra)
.IP \(bu 2
\f[CR]markdown_strict\f[R] (original unextended Markdown)
.IP \(bu 2
\f[CR]mediawiki\f[R] (MediaWiki markup)
.IP \(bu 2
\f[CR]man\f[R] (roff man)
.IP \(bu 2
\f[CR]mdoc\f[R] (mdoc manual page markup)
.IP \(bu 2
\f[CR]muse\f[R] (Muse)
.IP \(bu 2
\f[CR]native\f[R] (native Haskell)
.IP \(bu 2
\f[CR]odt\f[R] (OpenDocument text document)
.IP \(bu 2
\f[CR]opml\f[R] (OPML)
.IP \(bu 2
\f[CR]org\f[R] (Emacs Org mode)
.IP \(bu 2
\f[CR]pod\f[R] (Perl\(cqs Plain Old Documentation)
.IP \(bu 2
\f[CR]pptx\f[R] (PowerPoint)
.IP \(bu 2
\f[CR]ris\f[R] (RIS bibliography)
.IP \(bu 2
\f[CR]rtf\f[R] (Rich Text Format)
.IP \(bu 2
\f[CR]rst\f[R] (reStructuredText)
.IP \(bu 2
\f[CR]t2t\f[R] (txt2tags)
.IP \(bu 2
\f[CR]textile\f[R] (Textile)
.IP \(bu 2
\f[CR]tikiwiki\f[R] (TikiWiki markup)
.IP \(bu 2
\f[CR]twiki\f[R] (TWiki markup)
.IP \(bu 2
\f[CR]typst\f[R] (typst)
.IP \(bu 2
\f[CR]vimwiki\f[R] (Vimwiki)
.IP \(bu 2
\f[CR]xlsx\f[R] (Excel spreadsheet)
.IP \(bu 2
\f[CR]xml\f[R] (XML version of native AST)
.IP \(bu 2
the path of a custom Lua reader, see Custom readers and writers below
.PP
Extensions can be individually enabled or disabled by appending
\f[CR]+EXTENSION\f[R] or \f[CR]\-EXTENSION\f[R] to the format name.
See Extensions below, for a list of extensions and their names.
See \f[CR]\-\-list\-input\-formats\f[R] and
\f[CR]\-\-list\-extensions\f[R], below.
.RE
.TP
\f[CR]\-t\f[R] \f[I]FORMAT\f[R], \f[CR]\-w\f[R] \f[I]FORMAT\f[R], \f[CR]\-\-to=\f[R]\f[I]FORMAT\f[R], \f[CR]\-\-write=\f[R]\f[I]FORMAT\f[R]
Specify output format.
\f[I]FORMAT\f[R] can be:
.RS
.IP \(bu 2
\f[CR]ansi\f[R] (text with ANSI escape codes, for terminal viewing)
.IP \(bu 2
\f[CR]asciidoc\f[R] (modern AsciiDoc as interpreted by AsciiDoctor)
.IP \(bu 2
\f[CR]asciidoc_legacy\f[R] (AsciiDoc as interpreted by
\f[CR]asciidoc\-py\f[R]).
.IP \(bu 2
\f[CR]asciidoctor\f[R] (deprecated synonym for \f[CR]asciidoc\f[R])
.IP \(bu 2
\f[CR]bbcode\f[R] BBCode
.IP \(bu 2
\f[CR]bbcode_fluxbb\f[R] BBCode (FluxBB)
.IP \(bu 2
\f[CR]bbcode_phpbb\f[R] BBCode (phpBB)
.IP \(bu 2
\f[CR]bbcode_steam\f[R] BBCode (Steam)
.IP \(bu 2
\f[CR]bbcode_hubzilla\f[R] BBCode (Hubzilla)
.IP \(bu 2
\f[CR]bbcode_xenforo\f[R] BBCode (xenForo)
.IP \(bu 2
\f[CR]beamer\f[R] (LaTeX beamer slide show)
.IP \(bu 2
\f[CR]bibtex\f[R] (BibTeX bibliography)
.IP \(bu 2
\f[CR]biblatex\f[R] (BibLaTeX bibliography)
.IP \(bu 2
\f[CR]chunkedhtml\f[R] (zip archive of multiple linked HTML files)
.IP \(bu 2
\f[CR]commonmark\f[R] (CommonMark Markdown)
.IP \(bu 2
\f[CR]commonmark_x\f[R] (CommonMark Markdown with extensions)
.IP \(bu 2
\f[CR]context\f[R] (ConTeXt)
.IP \(bu 2
\f[CR]csljson\f[R] (CSL JSON bibliography)
.IP \(bu 2
\f[CR]djot\f[R] (Djot markup)
.IP \(bu 2
\f[CR]docbook\f[R] or \f[CR]docbook4\f[R] (DocBook 4)
.IP \(bu 2
\f[CR]docbook5\f[R] (DocBook 5)
.IP \(bu 2
\f[CR]docx\f[R] (Word docx)
.IP \(bu 2
\f[CR]dokuwiki\f[R] (DokuWiki markup)
.IP \(bu 2
\f[CR]epub\f[R] or \f[CR]epub3\f[R] (EPUB v3 book)
.IP \(bu 2
\f[CR]epub2\f[R] (EPUB v2)
.IP \(bu 2
\f[CR]fb2\f[R] (FictionBook2 e\-book)
.IP \(bu 2
\f[CR]gfm\f[R] (GitHub\-Flavored Markdown), or the deprecated and less
accurate \f[CR]markdown_github\f[R]; use \f[CR]markdown_github\f[R] only
if you need extensions not supported in \f[CR]gfm\f[R].
.IP \(bu 2
\f[CR]haddock\f[R] (Haddock markup)
.IP \(bu 2
\f[CR]html\f[R] or \f[CR]html5\f[R] (HTML, i.e.\ HTML5/XHTML polyglot
markup)
.IP \(bu 2
\f[CR]html4\f[R] (XHTML 1.0 Transitional)
.IP \(bu 2
\f[CR]icml\f[R] (InDesign ICML)
.IP \(bu 2
\f[CR]ipynb\f[R] (Jupyter notebook)
.IP \(bu 2
\f[CR]jats_archiving\f[R] (JATS XML, Archiving and Interchange Tag Set)
.IP \(bu 2
\f[CR]jats_articleauthoring\f[R] (JATS XML, Article Authoring Tag Set)
.IP \(bu 2
\f[CR]jats_publishing\f[R] (JATS XML, Journal Publishing Tag Set)
.IP \(bu 2
\f[CR]jats\f[R] (alias for \f[CR]jats_archiving\f[R])
.IP \(bu 2
\f[CR]jira\f[R] (Jira/Confluence wiki markup)
.IP \(bu 2
\f[CR]json\f[R] (JSON version of native AST)
.IP \(bu 2
\f[CR]latex\f[R] (LaTeX)
.IP \(bu 2
\f[CR]man\f[R] (roff man)
.IP \(bu 2
\f[CR]markdown\f[R] (Pandoc\(cqs Markdown)
.IP \(bu 2
\f[CR]markdown_mmd\f[R] (MultiMarkdown)
.IP \(bu 2
\f[CR]markdown_phpextra\f[R] (PHP Markdown Extra)
.IP \(bu 2
\f[CR]markdown_strict\f[R] (original unextended Markdown)
.IP \(bu 2
\f[CR]markua\f[R] (Markua)
.IP \(bu 2
\f[CR]mediawiki\f[R] (MediaWiki markup)
.IP \(bu 2
\f[CR]ms\f[R] (roff ms)
.IP \(bu 2
\f[CR]muse\f[R] (Muse)
.IP \(bu 2
\f[CR]native\f[R] (native Haskell)
.IP \(bu 2
\f[CR]odt\f[R] (OpenDocument text document)
.IP \(bu 2
\f[CR]opml\f[R] (OPML)
.IP \(bu 2
\f[CR]opendocument\f[R] (OpenDocument XML)
.IP \(bu 2
\f[CR]org\f[R] (Emacs Org mode)
.IP \(bu 2
\f[CR]pdf\f[R] (PDF)
.IP \(bu 2
\f[CR]plain\f[R] (plain text)
.IP \(bu 2
\f[CR]pptx\f[R] (PowerPoint slide show)
.IP \(bu 2
\f[CR]rst\f[R] (reStructuredText)
.IP \(bu 2
\f[CR]rtf\f[R] (Rich Text Format)
.IP \(bu 2
\f[CR]texinfo\f[R] (GNU Texinfo)
.IP \(bu 2
\f[CR]textile\f[R] (Textile)
.IP \(bu 2
\f[CR]slideous\f[R] (Slideous HTML and JavaScript slide show)
.IP \(bu 2
\f[CR]slidy\f[R] (Slidy HTML and JavaScript slide show)
.IP \(bu 2
\f[CR]dzslides\f[R] (DZSlides HTML5 + JavaScript slide show)
.IP \(bu 2
\f[CR]revealjs\f[R] (reveal.js HTML5 + JavaScript slide show)
.IP \(bu 2
\f[CR]s5\f[R] (S5 HTML and JavaScript slide show)
.IP \(bu 2
\f[CR]tei\f[R] (TEI Simple)
.IP \(bu 2
\f[CR]typst\f[R] (typst)
.IP \(bu 2
\f[CR]vimdoc\f[R] (Vimdoc)
.IP \(bu 2
\f[CR]xml\f[R] (XML version of native AST)
.IP \(bu 2
\f[CR]xwiki\f[R] (XWiki markup)
.IP \(bu 2
\f[CR]zimwiki\f[R] (ZimWiki markup)
.IP \(bu 2
the path of a custom Lua writer, see Custom readers and writers below
.PP
Note that \f[CR]odt\f[R], \f[CR]docx\f[R], \f[CR]epub\f[R], and
\f[CR]pdf\f[R] output will not be directed to \f[I]stdout\f[R] unless
forced with \f[CR]\-o \-\f[R].
.PP
Extensions can be individually enabled or disabled by appending
\f[CR]+EXTENSION\f[R] or \f[CR]\-EXTENSION\f[R] to the format name.
See Extensions below, for a list of extensions and their names.
See \f[CR]\-\-list\-output\-formats\f[R] and
\f[CR]\-\-list\-extensions\f[R], below.
.RE
.TP
\f[CR]\-o\f[R] \f[I]FILE\f[R], \f[CR]\-\-output=\f[R]\f[I]FILE\f[R]
Write output to \f[I]FILE\f[R] instead of \f[I]stdout\f[R].
If \f[I]FILE\f[R] is \f[CR]\-\f[R], output will go to \f[I]stdout\f[R],
even if a non\-textual format (\f[CR]docx\f[R], \f[CR]odt\f[R],
\f[CR]epub2\f[R], \f[CR]epub3\f[R]) is specified.
If the output format is \f[CR]chunkedhtml\f[R] and \f[I]FILE\f[R] has no
extension, then instead of producing a \f[CR].zip\f[R] file pandoc will
create a directory \f[I]FILE\f[R] and unpack the zip archive there
(unless \f[I]FILE\f[R] already exists, in which case an error will be
raised).
.TP
\f[CR]\-\-data\-dir=\f[R]\f[I]DIRECTORY\f[R]
Specify the user data directory to search for pandoc data files.
If this option is not specified, the default user data directory will be
used.
On *nix and macOS systems this will be the \f[CR]pandoc\f[R]
subdirectory of the XDG data directory (by default,
\f[CR]$HOME/.local/share\f[R], overridable by setting the
\f[CR]XDG_DATA_HOME\f[R] environment variable).
If that directory does not exist and \f[CR]$HOME/.pandoc\f[R] exists, it
will be used (for backwards compatibility).
On Windows the default user data directory is
\f[CR]%APPDATA%\(rspandoc\f[R].
You can find the default user data directory on your system by looking
at the output of \f[CR]pandoc \-\-version\f[R].
Data files placed in this directory (for example,
\f[CR]reference.odt\f[R], \f[CR]reference.docx\f[R],
\f[CR]epub.css\f[R], \f[CR]templates\f[R]) will override pandoc\(cqs
normal defaults.
(Note that the user data directory is not created by pandoc, so you will
need to create it yourself if you want to make use of it.)
.TP
\f[CR]\-d\f[R] \f[I]FILE\f[R], \f[CR]\-\-defaults=\f[R]\f[I]FILE\f[R]
Specify a set of default option settings.
\f[I]FILE\f[R] is a YAML or JSON file whose fields correspond to
command\-line option settings.
All options for document conversion, including input and output files,
can be set using a defaults file.
The file will be searched for first in the working directory, and then
in the \f[CR]defaults\f[R] subdirectory of the user data directory (see
\f[CR]\-\-data\-dir\f[R]).
The \f[CR].yaml\f[R] extension will be added if \f[I]FILE\f[R] lacs an
extension.
See the section Defaults files for more information on the file format.
Settings from the defaults file may be overridden or extended by
subsequent options on the command line.
.TP
\f[CR]\-\-bash\-completion\f[R]
Generate a bash completion script.
To enable bash completion with pandoc, add this to your
\f[CR].bashrc\f[R]:
.RS
.IP
.EX
eval \(dq$(pandoc \-\-bash\-completion)\(dq
.EE
.RE
.TP
\f[CR]\-\-verbose\f[R]
Give verbose debugging output.
.TP
\f[CR]\-\-quiet\f[R]
Suppress warning messages.
.TP
\f[CR]\-\-fail\-if\-warnings[=true|false]\f[R]
Exit with error status if there are any warnings.
.TP
\f[CR]\-\-log=\f[R]\f[I]FILE\f[R]
Write log messages in machine\-readable JSON format to \f[I]FILE\f[R].
All messages above DEBUG level will be written, regardless of verbosity
settings (\f[CR]\-\-verbose\f[R], \f[CR]\-\-quiet\f[R]).
.TP
\f[CR]\-\-list\-input\-formats\f[R]
List supported input formats, one per line.
.TP
\f[CR]\-\-list\-output\-formats\f[R]
List supported output formats, one per line.
.TP
\f[CR]\-\-list\-extensions\f[R][\f[CR]=\f[R]\f[I]FORMAT\f[R]]
List supported extensions for \f[I]FORMAT\f[R], one per line, preceded
by a \f[CR]+\f[R] or \f[CR]\-\f[R] indicating whether it is enabled by
default in \f[I]FORMAT\f[R].
If \f[I]FORMAT\f[R] is not specified, defaults for pandoc\(cqs Markdown
are given.
.TP
\f[CR]\-\-list\-highlight\-languages\f[R]
List supported languages for syntax highlighting, one per line.
.TP
\f[CR]\-\-list\-highlight\-styles\f[R]
List supported styles for syntax highlighting, one per line.
See \f[CR]\-\-syntax\-highlighting\f[R].
.TP
\f[CR]\-v\f[R], \f[CR]\-\-version\f[R]
Print version.
.TP
\f[CR]\-h\f[R], \f[CR]\-\-help\f[R]
Show usage message.
.SS Reader options
.TP
\f[CR]\-\-shift\-heading\-level\-by=\f[R]\f[I]NUMBER\f[R]
Shift heading levels by a positive or negative integer.
For example, with \f[CR]\-\-shift\-heading\-level\-by=\-1\f[R], level 2
headings become level 1 headings, and level 3 headings become level 2
headings.
Headings cannot have a level less than 1, so a heading that would be
shifted below level 1 becomes a regular paragraph.
Exception: with a shift of \-N, a level\-N heading at the beginning of
the document replaces the metadata title.
\f[CR]\-\-shift\-heading\-level\-by=\-1\f[R] is a good choice when
converting HTML or Markdown documents that use an initial level\-1
heading for the document title and level\-2+ headings for sections.
\f[CR]\-\-shift\-heading\-level\-by=1\f[R] may be a good choice for
converting Markdown documents that use level\-1 headings for sections to
HTML, since pandoc uses a level\-1 heading to render the document title.
.TP
\f[CR]\-\-base\-header\-level=\f[R]\f[I]NUMBER\f[R]
\f[I]Deprecated.
Use \f[CI]\-\-shift\-heading\-level\-by\f[I]=X instead, where X = NUMBER
\- 1.\f[R] Specify the base level for headings (defaults to 1).
.TP
\f[CR]\-\-indented\-code\-classes=\f[R]\f[I]CLASSES\f[R]
Specify classes to use for indented code blocks\(emfor example,
\f[CR]perl,numberLines\f[R] or \f[CR]haskell\f[R].
Multiple classes may be separated by spaces or commas.
.TP
\f[CR]\-\-default\-image\-extension=\f[R]\f[I]EXTENSION\f[R]
Specify a default extension to use when image paths/URLs have no
extension.
This allows you to use the same source for formats that require
different kinds of images.
Currently this option only affects the Markdown and LaTeX readers.
.TP
\f[CR]\-\-file\-scope[=true|false]\f[R]
Parse each file individually before combining for multifile documents.
This will allow footnotes in different files with the same identifiers
to work as expected.
If this option is set, footnotes and links will not work across files.
Reading binary files (docx, odt, epub) implies
\f[CR]\-\-file\-scope\f[R].
.RS
.PP
If two or more files are processed using \f[CR]\-\-file\-scope\f[R],
prefixes based on the filenames will be added to identifiers in order to
disambiguate them, and internal links will be adjusted accordingly.

04_labs/corpus/external_frameworks/pandoc/wasm/pandoc.js
/* pandoc.js: JavaScript interface to pandoc.wasm.
   Copyright (c) 2025 Tweag I/O Limited and John MacFarlane. MIT License.

   Interface:

   await convert(options, stdin, files)

   - options is a JavaScript object representing pandoc options: this should
     correspond to the format used in pandoc's default files.
   - stdin is a string or nil
   - files is a JavaScript object whose keys are filenames and whose values
     are the data in the corresponding file, as Blobs.

   The return value is a JavaScript object with 3 properties, stdout,
   stderr, and warnings, all strings. warnings is a JSON-encoded
   version of the warnings produced by pandoc. If the pandoc process
   produces an output file, it will be added to files.

   await query(options)

    - options is a JavaScript object with a 'query' property and in
      some cases a 'format' property. Possible queries include
      'version', 'highlight-styles', 'highlight-languages', 'input-formats',
      'output-formats', 'default-template' (requires 'format'),
      and 'extensions-for-format' (requires 'format').

   The return value is a JavaScript string or in some cases a list
   of strings.
*/

import {
  WASI,
  OpenFile,
  File,
  ConsoleStdout,
  PreopenDirectory,
} from "https://cdn.jsdelivr.net/npm/@bjorn3/browser_wasi_shim@0.3.0/dist/index.js";

const args = ["pandoc.wasm", "+RTS", "-H64m", "-RTS"];
const env = [];
const in_file = new File(new Uint8Array(), { readonly: true });
const out_file = new File(new Uint8Array(), { readonly: false });
const err_file = new File(new Uint8Array(), { readonly: false });
const warnings_file = new File(new Uint8Array(), { readonly: false });
const fileSystem = new Map();
const fds = [
  new OpenFile(new File(new Uint8Array(), { readonly: true })),
  ConsoleStdout.lineBuffered((msg) => console.log(`[WASI stdout] ${msg}`)),
  ConsoleStdout.lineBuffered((msg) => console.warn(`[WASI stderr] ${msg}`)),
  new PreopenDirectory("/", fileSystem),
];
const options = { debug: false };
const wasi = new WASI(args, env, fds, options);
const { instance } = await WebAssembly.instantiateStreaming(
  fetch("./pandoc.wasm?sha1=SHA1_PANDOC_WASM"),
  {
    wasi_snapshot_preview1: wasi.wasiImport,
  }
);

wasi.initialize(instance);
instance.exports.__wasm_call_ctors();

function memory_data_view() {
  return new DataView(instance.exports.memory.buffer);
}

const argc_ptr = instance.exports.malloc(4);
memory_data_view().setUint32(argc_ptr, args.length, true);
const argv = instance.exports.malloc(4 * (args.length + 1));
for (let i = 0; i < args.length; ++i) {
  const arg = instance.exports.malloc(args[i].length + 1);
  new TextEncoder().encodeInto(
    args[i],
    new Uint8Array(instance.exports.memory.buffer, arg, args[i].length)
  );
  memory_data_view().setUint8(arg + args[i].length, 0);
  memory_data_view().setUint32(argv + 4 * i, arg, true);
}
memory_data_view().setUint32(argv + 4 * args.length, 0, true);
const argv_ptr = instance.exports.malloc(4);
memory_data_view().setUint32(argv_ptr, argv, true);

instance.exports.hs_init_with_rtsopts(argc_ptr, argv_ptr);

export async function query(options) {
  const opts_str = JSON.stringify(options);
  const opts_bytes = new TextEncoder().encode(opts_str);
  const opts_ptr = instance.exports.malloc(opts_bytes.length);
  new Uint8Array(instance.exports.memory.buffer, opts_ptr, opts_bytes.length)
    .set(opts_bytes);
  // add input files to fileSystem
  fileSystem.clear()
  const out_file = new File(new Uint8Array(), { readonly: false });
  const err_file = new File(new Uint8Array(), { readonly: false });
  fileSystem.set("stdout", out_file);
  fileSystem.set("stderr", err_file);
  instance.exports.query(opts_ptr, opts_bytes.length);

  const err_text = new TextDecoder("utf-8", { fatal: true }).decode(err_file.data);
  if (err_text) console.log(err_text);
  const out_text = new TextDecoder("utf-8", { fatal: true }).decode(out_file.data);
  return JSON.parse(out_text);
}


export async function convert(options, stdin, files) {
  const opts_str = JSON.stringify(options);
  const opts_bytes = new TextEncoder().encode(opts_str);
  const opts_ptr = instance.exports.malloc(opts_bytes.length);
  new Uint8Array(instance.exports.memory.buffer, opts_ptr, opts_bytes.length)
    .set(opts_bytes);
  // add input files to fileSystem
  fileSystem.clear()
  const in_file = new File(new Uint8Array(), { readonly: true });
  const out_file = new File(new Uint8Array(), { readonly: false });
  const err_file = new File(new Uint8Array(), { readonly: false });
  const warnings_file = new File(new Uint8Array(), { readonly: false });
  fileSystem.set("stdin", in_file);
  fileSystem.set("stdout", out_file);
  fileSystem.set("stderr", err_file);
  fileSystem.set("warnings", warnings_file);
  for (const file in files) {
    await addFile(file, files[file], true);
  }
  // add output file if any
  if (options["output-file"]) {
    await addFile(options["output-file"], new Blob(), false);
  }
  // add media file for extracted media
  if (options["extract-media"]) {
    await addFile(options["extract-media"], new Blob(), false);
  }
  if (stdin) {
    in_file.data = new TextEncoder().encode(stdin);
  }
  instance.exports.convert(opts_ptr, opts_bytes.length);

  if (options["output-file"]) {
    files[options["output-file"]] =
       new Blob([fileSystem.get(options["output-file"]).data]);
  }
  if (options["extract-media"]) {
    const mediaFile = fileSystem.get(options["extract-media"]);
    if (mediaFile && mediaFile.data && mediaFile.data.length > 0) {
      files[options["extract-media"]] =
         new Blob([mediaFile.data], { type: 'application/zip' });
    }
  }
  const rawWarnings = new TextDecoder("utf-8", { fatal: true })
                          .decode(warnings_file.data);
  let warnings = [];
  if (rawWarnings) {
    warnings = JSON.parse(rawWarnings);
  }
  return {
    stdout: new TextDecoder("utf-8", { fatal: true }).decode(out_file.data),
    stderr: new TextDecoder("utf-8", { fatal: true }).decode(err_file.data),
    warnings: warnings
  };
}

async function addFile(filename, blob, readonly) {
  const buffer = await blob.arrayBuffer();
  const file = new File(new Uint8Array(buffer), { readonly: readonly });
  fileSystem.set(filename, file);
}
